net/tcp/tcp_recvwindow.c

```diff
@@ -163,3 +163,80 @@ uint16_t tcp_get_recvwindow(FAR struct net_driver_s *dev,
 
   return recvwndo;
 }
+
+bool tcp_should_send_recvwindow(FAR struct tcp_conn_s *conn)
+{
+  FAR struct net_driver_s *dev = conn->dev;
+  uint16_t win;
+  size_t maxiob;
+  uint16_t maxwin;
+  uint16_t oldwin;
+  uint32_t rcvseq;
+  uint16_t adv;
+  uint16_t mss;
+
+  /* If the window shrunk, don't send. */
+
+  rcvseq = tcp_getsequence(conn->rcvseq);
+  if (TCP_SEQ_GTE(conn->rcv_adv, rcvseq))
+    {
+      oldwin = TCP_SEQ_SUB(conn->rcv_adv, rcvseq);
+    }
+  else
+    {
+      /* This can happen for example when:
+       *
+       * - we shrunk the window
+       * - zero window probes advanced rcvseq
+       */
+
+      oldwin = 0;
+    }
+
+  win = tcp_get_recvwindow(dev, conn);
+  if (win <= oldwin)
+    {
+      return false;
+    }
+
+  adv = win - oldwin;
+
+  /* The following conditions are inspired from NetBSD TCP stack.
+   *
+   * - If we can extend the window by the half of the max possible size,
+   *   send it.
+   *
+   * - If we can extend the window by 2 * mss, send it.
+   */
+
+  /* Calculate the max possible window size for the connection.
+   * This needs to be in sync with tcp_get_recvwindow().
+   */
+
+  maxiob = (CONFIG_IOB_NBUFFERS - CONFIG_IOB_THROTTLE) * CONFIG_IOB_BUFSIZE;
+  if (maxiob >= UINT16_MAX)
+    {
+      maxwin = UINT16_MAX;
+    }
+  else
+    {
+      maxwin = maxiob;
+    }
```

Comment:
should we add a new function for line 212-224 and let here and tcp_get_recvwindow call it?

Reply:
do you mean the maxwin calculation? i will.

Reply:
Yes.

Reply:
done